### PR TITLE
Migrate Scala3doc repo to Dotty

### PR DESCRIPTION
**Pull request by romanowski**

### CODE REVIEW

File: .vscode/settings.json
Comment (abgruszecki) [inline .vscode/settings.json:1]: This file should probably be removed.

**Review by smarter — CHANGES_REQUESTED** [review]
I'll look at the build more later but from a quick look:

- scala3doc has its own build.sbt, it's the only project doing that so I'd rather avoid that and merge that in the main Build.scala
- There's a scala3doc and a scala3doc-test project, again for consistency with everything else it'd be better to have a single project and use the test configuration of the main project to run the tests (we have way too many projects in our build already).

**Comment by romanowski** [comment]
> scala3doc has its own build.sbt, it's the only project doing that so I'd rather avoid that and merge that in the main Build.scala
There are two reasons why I did it this way:
- `Build.scala` is huge and I think it will be a good idea to gradually moving project-specific configurations and tasks to `build.sbt` files
- This PR adds rules not to run all compiler checks if PRs only touches `scala3doc*` projects and having as much configuration as possible whitn that directories will make our work on scala3doc much faster

@smarter is there any other reasong then convention that why we move project-specific configuration to `Build.scala`?

>  There's a scala3doc and a scala3doc-test project, again for consistency with everything else it'd be better to have a single project and use the test configuration of the main project to run the tests (we have way too many projects in our build already).

I will change that.

**Comment by smarter** [comment]
> Build.scala is huge and I think it will be a good idea to gradually moving project-specific configurations and tasks to build.sbt files

I think using multiple build.sbt will increase the sbt load times. The reason we have the whole build in one file is that @sjrd told me that was better than the alternative and he's the one with the most experience with this stuff so I defer to him on that :)

> This PR adds rules not to run all compiler checks if PRs only touches scala3doc* projects and having as much configuration as possible whitn that directories will make our work on scala3doc much faster

You mean run as part of the CI? If you're touching any part of the build I'd like to have the full CI run before merging the PR anyway because it could have consequences on something else.

**Comment by romanowski** [comment]
@smarter I've moved configuration to `Build.scala` since I really want to merge this PR as soon as possible (fixing conflicts with other PRs are real pain).

> You mean run as part of the CI? If you're touching any part of the build I'd like to have the full CI run before merging the PR anyway because it could have consequences on something else.

The problem with running everything is I faced today a lot of cases where jobs was aborted or canceled. Brigingin scala3doc on board will add 3 more developers that will work on daily basis on code so we may see this problem more and more.

**Comment by smarter** [comment]
> The problem with running everything is I faced today a lot of cases where jobs was aborted or canceled. 

Yeah the CI is not working great right now, but there's things we can try to improve that (stop running the non-bootstrapped job on PR, tweak the parallelism, etc), I agree it's important to get it back to working reliably (it's only been this bad recently, not sure why exactly)

**Comment by romanowski** [comment]
@smarter is there anything else I should fix? Or this PR looks good from your perspective?

**Comment by smarter** [comment]
I'll have a more detailed look at the build tomorrow.

**Comment by smarter** [comment]
I investigated and found that one of our CI runner had been stuck all day not processing request and that's why everything was broken, I restarted it and things seem to be working now, sorry that you ended up working on scala3doc just on this day!

**Comment by smarter** [comment]
(I restarted a lot of jobs which didn't run today, so you'll have to wait a while before this PR gets run)

**Comment by romanowski** [comment]
@MasseGuillaume it seems that CI is misbehaving again today. Previous run started however compile step does not start in almost 20 minutes.

**Comment by smarter** [comment]
Wrong nick :). As far as I can tell it's not misbehaving, from the logs in https://github.com/lampepfl/dotty/runs/1320094163?check_suite_focus=true:
> 2020-10-28T10:32:36.5547756Z Found online and busy self-hosted runners in current repository that matches the required labels: 'self-hosted , Linux'. Waiting for one of them to get assigned for this job.

So it was just waiting for a runner to be free (which makes sense, many other PRs are being tested at the same time: https://github.com/lampepfl/dotty/actions), imho it's a bug in github actions if the timeout starts before it has found a free runner.

**Comment by romanowski** [comment]
Nick fixed, sorry. Yeah, I've switch from hostem job to public one and it is the onle one completed so far.

File: project/Build.scala
Comment (smarter) [inline project/Build.scala:283]: Looks like commonBootstrappedSettings0 isn't used anymore so this change isn't needed.

File: project/Build.scala
Comment (smarter) [inline project/Build.scala:1489]: This is already set in `thisBuildSettings` so shouldn't needed be needed.

File: project/Build.scala
Comment (smarter) [inline project/Build.scala:1493]: This is already set in `thisBuildSettings` so shouldn't needed be needed.

File: project/Build.scala
Comment (smarter) [inline project/Build.scala:1486]: I think `test.in(Test)` is the same as `Test / test`, the latter is the recommended style now.

File: project/Build.scala
Comment (smarter) [inline project/Build.scala:1173]: If they're just testcases, why are they in their own project in their own directory? Could they be in `scala3doc/test-resources` instead? (just like we have `compiler/test-resources`)

File: project/Build.scala
Comment (smarter) [inline project/Build.scala:1474]: Could you add a comment explaining what we need each of these resolvers for? Is this only temporary?

File: project/Build.scala
Comment (smarter) [inline project/Build.scala:1481]: Is this really needed? We already have a way to process arguments (https://github.com/lampepfl/dotty/blob/d09600de025e6760dc8614a4ce62d568e90d16d5/compiler/src/dotty/tools/dotc/config/CompilerCommand.scala#L36-L37), we shouldn't use multiple different argument parsers, and we shouldn't depend on external projects unless absolutely necessary.

File: sbt-dotty/src/dotty/tools/sbtplugin/DottyPlugin.scala
Comment (smarter) [inline sbt-dotty/src/dotty/tools/sbtplugin/DottyPlugin.scala:368]: What is this for? Some documentation would be useful

File: project/Build.scala
Comment (smarter) [inline project/Build.scala:1171]: why is generateSelfDocumentation an input task but the other ones aren't?

File: project/Build.scala
Comment (smarter) [inline project/Build.scala:1476]: For internal dependency we should use `dependsOn` (and in fact there is ``dependsOn(`scala3-tasty-inspector`)`` above so this line can just be removed)

File: project/Build.scala
Comment (smarter) [inline project/Build.scala:1465]: scalac/scaladoc/dotc/dottydoc have always used `-d` for the output directory, not `-o`: https://github.com/lampepfl/dotty/blob/d09600de025e6760dc8614a4ce62d568e90d16d5/compiler/src/dotty/tools/dotc/config/ScalaSettings.scala#L24

File: .github/workflows/scala3doc.yaml
Comment (smarter) [inline .github/workflows/scala3doc.yaml:58]: Isn't this basically equivalent to running the sbt scripted test [`sbt-dotty/scala3doc`](https://github.com/romanowski/dotty/tree/scala3doc/migration/sbt-dotty/sbt-test/sbt-dotty/scala3doc)?
```scala
sbt "sbt-dotty/scripted sbt-dotty/scala3doc"
```

**Review by smarter — CHANGES_REQUESTED** [review]

File: .github/workflows/scala3doc.yaml
Comment (romanowski) [reply .github/workflows/scala3doc.yaml:58]: Yes, the only difference is that the project is bigger but I can move it there.

**Comment by bishabosha** [comment]
I just merged #10061

File: project/Build.scala
Comment (romanowski) [reply project/Build.scala:1173]: We want to compile them since we are using generated .tasty files in our tests

File: sbt-dotty/src/dotty/tools/sbtplugin/DottyPlugin.scala
Comment (romanowski) [reply sbt-dotty/src/dotty/tools/sbtplugin/DottyPlugin.scala:368]: I've added a comment here and created a issue for proper documentation: https://github.com/lampepfl/scala3doc/issues/222

File: project/Build.scala
Comment (romanowski) [reply project/Build.scala:1171]: Good catch it can be a plain task

File: project/Build.scala
Comment (romanowski) [reply project/Build.scala:1476]: ah, we've forgot to remove that one during migration, will fix that

File: project/Build.scala
Comment (romanowski) [reply project/Build.scala:1474]: sure

File: project/Build.scala
Comment (smarter) [reply project/Build.scala:1175]: OK, so could `generateTestcasesDocumentation` be replaced by `scala3doc-testcases/doc` or is there a difference between the two?

**Comment by romanowski** [comment]
> Any reason dokka isn't on maven central? Can we expect it to be on maven central at some point? Having to add custom resolvers in user builds is really not great.

It should be publish to mavenCentral soon up to my knowladge.

**Comment by smarter** [comment]
> It should be publish to mavenCentral soon up to my knowladge.

OK, it'd be good to have an issue to keep track of that.

File: project/Build.scala
Comment (romanowski) [reply project/Build.scala]: I've tried to do that here: https://github.com/romanowski/dotty/commit/b9ac86d813b053915bd5b06dff283c7b1574cdd6 however there are few problems:
- inside compiler all arguments seems to be optional, so I am forced to wite pice of code to enforce that
- aliases exists in API however it seems that are not implemented

Since this is non-trivial task (especially that we will need to add test and refator a bit our API) maybe we can do it after initial merge (#10091 broke this PR and I will need to recreate this PR again)? I've created issue to track that: https://github.com/lampepfl/scala3doc/issues/223

Comment (smarter) [reply project/Build.scala]: > Since this is non-trivial task (especially that we will need to add test and refator a bit our API) maybe we can do it after initial merge

Yes, that's fine.

File: project/Build.scala
Comment (romanowski) [reply project/Build.scala:1171]: For now `scala3doc-testcases/doc` will not generate any documentation since we are not using (yet) scala3doc in any of our projects directly. This will involve working around cyclic depenencies and we can look at it after this PR is merged.

**Comment by romanowski** [comment]
I create https://github.com/lampepfl/scala3doc/issues/226 to keep track of this cutsom resolvers.

File: project/Build.scala
Comment (abgruszecki) [reply project/Build.scala:1478]: Correct me if I'm wrong, but doesn't this mean that any change to our CLI would require changing the compiler's CLI? Because I think that's the case, and I think that we should have a separate arg parser for the doctool.

File: sbt-dotty/src/dotty/tools/sbtplugin/DottyPlugin.scala
Comment (abgruszecki) [reply sbt-dotty/src/dotty/tools/sbtplugin/DottyPlugin.scala:368]: So the reason we need this is that we have a separate CLI interface with its own arguments, and we may need to forward options to it (that's what `scala3docOptions` keeps track of). When we receive options from SBT, we strip off the ones starting with `--+DOC+`, and pass these along with some default values we extracted from compiler args to the doctool arg parser.

**Pull request merged**
